internal/clients/kibana/alerting.go

```diff
@@ -0,0 +1,181 @@
+package kibana
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/elastic/terraform-provider-elasticstack/generated/alerting"
+	"github.com/elastic/terraform-provider-elasticstack/internal/clients"
+	"github.com/elastic/terraform-provider-elasticstack/internal/models"
+	"github.com/elastic/terraform-provider-elasticstack/internal/utils"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+)
+
+func ruleResponseToModel(spaceID string, res *alerting.RuleResponseProperties) *models.AlertingRule {
+	if res == nil {
+		return nil
+	}
+
+	actions := []models.AlertingRuleAction{}
+	for _, action := range res.Actions {
+		actions = append(actions, models.AlertingRuleAction{
+			Group:  *action.Group,
+			ID:     *action.Id,
+			Params: action.Params,
+		})
+	}
+
+	scheduledTaskId := ""
+	if res.ScheduledTaskId != nil {
+		scheduledTaskId = *res.ScheduledTaskId
+	}
+
+	return &models.AlertingRule{
+		ID:         res.Id,
+		SpaceID:    spaceID,
+		Name:       res.Name,
+		Consumer:   res.Consumer,
+		NotifyWhen: string(*res.NotifyWhen),
+		Params:     res.Params,
+		RuleTypeID: res.RuleTypeId,
+		Schedule: models.AlertingRuleSchedule{
+			Interval: *res.Schedule.Interval,
+		},
+		Enabled:         &res.Enabled,
+		Tags:            res.Tags,
+		Throttle:        res.Throttle.Get(),
+		ScheduledTaskID: scheduledTaskId,
+		ExecutionStatus: models.AlertingRuleExecutionStatus{
+			LastExecutionDate: res.ExecutionStatus.LastExecutionDate.String(),
+			Status:            *res.ExecutionStatus.Status,
+		},
+		Actions: actions,
+	}
+}
+
+func PutAlertingRule(ctx context.Context, apiClient *clients.ApiClient, isCreate bool, rule models.AlertingRule) (*models.AlertingRule, diag.Diagnostics) {
+	client, err := apiClient.GetAlertingClient()
+	if err != nil {
+		return nil, diag.FromErr(err)
+	}
+
+	ctxWithAuth := apiClient.SetAlertingAuthContext(ctx)
+
+	actions := []alerting.ActionsInner{}
+	for _, action := range rule.Actions {
+		actions = append(actions, alerting.ActionsInner{
+			Group:  &action.Group,
+			Id:     &action.ID,
+			Params: action.Params,
+		})
+	}
+
+	if isCreate {
```

Comment:
Does it make sense to split the function into Create and Update ones? It looks like the only common part for these operations is the lines #57 - #71 that can be moved into a helper func. IMO upsert functions tend to be more convoluted and error prone than a pair of create and update.